A better SponsorBlock

nihil-admirari opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

Changes are similar in purpose to [SponSkrub](https://github.com/faissaloo/SponSkrub), but the implementation is completely different.

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This is a complete replacement of SponSkrub. New features:
1. Proper cut implemented using [`concat` FFmpeg demuxer](https://ffmpeg.org/ffmpeg-formats.html#concat-1), mentioned in https://github.com/faissaloo/SponSkrub/issues/32. According to https://sattlers.org/2019/09/30/ffmpeg-split-and-cut-video-segments/, small stutter/frame skipping mentioned in the issue can be fixed by forcing key frames, an option for which is also provided.
2. Implementation takes `videoDuration` from SponsorBlock responses into account.
3. Implementation is compatible with `--skip-chapters`.
4. Different chapter merge algorithm, which does not lead to "Chapter end time X before chapter start Y" reported in https://github.com/faissaloo/SponSkrub/issues/27.
5. SRT, ASS, and VTT subtitles (seems like [FFmpeg does not fully support others](https://ffmpeg.org/general.html#Subtitle-Formats)) are cut along with the video, solving https://github.com/faissaloo/SponSkrub/issues/21.
6. Fully cross-platform (SponSkrub still doesn't have macOS binaries).
7. Requires no external dependencies.
8. Written in Python, not in D, which makes contributing much easier.

pukkandan commented:
I was also planning to rewrite the sponskrub integration. However, my plan was quite different from your implementation. Here is what I was thinking:
1. Don't touch sponskrub. Implement this independantly
1. Implement chapter merger as a utility and sponsorblock API to youtube extractor itself. This automatically takes care of the chapter mode. User can just do `--add-metadata --use-sponsorblock`. This also makes it so that the sponsorblock data is available in the info-json if needed
2. Write a concat postprocessor since it has other uses as well. See https://github.com/ytdl-org/youtube-dl/pull/28153 and related issues
3. Add an option `--remove-chapters` that internally uses `SplitChaptersPP` and the `concatPP` to remove any chapter from the video. Now, using `--remove-chapters sponsorblock.+` will take care of the sponskrub cut mode. This option automatically also allows user to specify the categories to be cut

Doing it this way would provide us with a lot more flexibility imo. Let me know what you think of this appoach

pukkandan commented:
> Add an option --remove-chapters that internally uses SplitChaptersPP and the concatPP to remove any chapter 

Looking at your code in more detail, using your method of `inpoint`/`outpoint` is indeed a better approach than splitting and then concatting

nihil-admirari commented:
1. Should I bring the entire SponSkrub implementation back, disable it by default and hide its options from help, or just create aliases `--sponskrub -> --sponsorblock` etc.?
2. Looks like SponsorBlock has [plans to support other sites](https://github.com/ajayyy/SponsorBlock/issues/515); server-side already accepts `service` query parameter, so moving SponsorBlock to YouTube extractor might not be a good idea.

    I'll look at updating info-json.
3. Concat is a very raw operation; I'm afraid, a generic `FFmpegConcatPostProcessor` will have to handle SponsorBlock, handle playlists, and over time maybe other operations, which have only concatenation in common, making it difficult to maintain.

    I can, however, move concat code to the base class `FFmpegPostProcessor`, so that if someone is going to implement `ConcatPlaylistPostProcessor`, he can reuse the implementation. 
4. `--remove-chapters sponsorblock.+` is not exactly what cut does. If there is an already existing `Chapter` with a `sponsor` in it, `SponsorBlock` will create chapters `Chapter - 1`, `Sponsor`, `Chapter - 2`. Cut however will simply decrease `Chapter`'s duration, without splitting it into `Chapter - 1`, `Chapter - 2`, which `--remove-chapters sponsorblock.+` implies.

    Specifying categories to cut will take some time to implement, chapter merge algorithms need to be updated to handle it. Handling categories to cut would make removing arbitrary chapters possible, but to implement `--remove-chapters`, I would rather suggest to extract `ChaptersPP` from `FFmpegMetadataPP` (SponsorBlock already has to pass flags into it), and moving SponsorBlock code into `ChaptersPP` to have all chapter handling code in one place. What do you think about it?

pukkandan commented:
> Should I bring the entire SponSkrub implementation back, disable it by default and hide its options from help, or just create aliases --sponskrub -> --sponsorblock etc.?

Yes, leave sponskrub as-is for now. We'll worry about how to document it once this new implementation is complete

> Looks like SponsorBlock has plans to support other sites; server-side already accepts service query parameter, so moving SponsorBlock to YouTube extractor might not be a good idea.

In that case, we should split the PP into `SponsorBlockPP` and `RemoveChaptersPP`. Then call the `SponsorBlockPP` as a pre-processor (like MetadataFromFieldPP)

> Concat is a very raw operation; I'm afraid, a generic FFmpegConcatPostProcessor will have to handle SponsorBlock, handle playlists, and over time maybe other operations, which have only concatenation in common, making it difficult to maintain. 

I am not familiar enough with ffmpeg to comment on this. I'll trust your judgement on this for now. If need be, we can generalize it later

> I can, however, move concat code to the base class FFmpegPostProcessor, so that if someone is going to implement ConcatPlaylistPostProcessor, he can reuse the implementation.

Yes, please do so if possible

> --remove-chapters sponsorblock.+ is not exactly what cut does. If there is an already existing Chapter with a sponsor in it, SponsorBlock will create chapters Chapter - 1, Sponsor, Chapter - 2. Cut however will simply decrease Chapter's duration, without splitting it into Chapter - 1, Chapter - 2, which --remove-chapters sponsorblock.+ implies.

It is though. Take this example: the video has chapters `A`, `B` with sponsor sections in the middle of `A`. The user calls yt-dlp with  something like `--sponsorblock --remove-chapters "sponsorblock-.+"`. So `SponsorBlock` preprocessor will first call the API and merge the chapter data to give `A-1`, `sponsorblock-sponsor`, `A-2`, `B`. This is saved in the infojson. Now `RemoveChaptersPP` will run and remove the chapters that match the regex causing the video to end up with the chapters `A-1`, `A-2`, `B`. Ofcourse, unless the user provides `--add-metadata`, this chapter info is not written to file

> Specifying categories to cut will take some time to implement, chapter merge algorithms need to be updated to handle it. 

I don't understand. Nothing needs to change in the chapter merger. It is simply that when we cut sections out, we cut off only the chapters matching the regex given by user.

> I would rather suggest to extract ChaptersPP from FFmpegMetadataPP (SponsorBlock already has to pass flags into it), and moving SponsorBlock code into ChaptersPP to have all chapter handling code in one place

If needed, we could seperate the chapter code into its own function. However, this is not necessary for this PR. Once we put the modified chapter data in the infodict using `SponsorBlock` pre-processor, `FFmpegMetadataPP` will automatically take care of adding these chapters

faissaloo commented on yt_dlp/postprocessor/sponsorblock.py:
```diff
@@ -0,0 +1,391 @@
+        # Segments returned by SponsorBlock often overlap;
+        # we need to merge segments before converting them into chapters.
+        overlap = False
+        for i in range(1, len(segments)):
+            s = segments[i]
+            start, end = s['segment']
+            if start <= cur_end:
+                overlap = True
+                cur_end, cat = max(cur_end, end), s['category']
+                cur_cats[cat] = CATEGORY_DISPLAY_NAMES[cat]
+            else:
+                append_chapter()
+                cur_start, cur_end, cur_cats = extract_start_end_cat(s)
+        append_chapter()
```
If I'm understanding this correctly this'll merge subsegments into one big segment under one category right? The reason I went with the approach I did was so that individual categories could be skipped independently, the user would have less control with your approach if I've understood it correctly.

nihil-admirari replied:
All merged categories are reflected in the title:
```py
title = '[SponsorBlock]: ' + '/'.join(cur_cats.values())
```
Resulting title will look like this: `[SponsorBlock]: Sponsor/Intro/Self-Promotion`.

I give user a warning when overlap is detected:
```py
'Some SponsorBlock segments overlap. '
'Usually it means that there are many conflicting submissions. '
'You may want to review them, and upvote or downvote appropriately'
```

The rationale for this can be seen e.g. here:
(https://sponsor.ajay.app/api/skipSegments?videoID=fFFSKedy9f4&categories=["sponsor","intro","outro","interaction","selfpromo","music_offtopic"]. Someone marked the entire beginning of the video as `sponsor`, which was later refined into separate `intro` and `selfpromo`. The code on its own cannot decide which submissions are correct, that's why the user is warned to review and to upvote/downvote. In this particular case, separate `intro` and `selfpromo` are better, but it may be the other way around for another video.

It is also very hard to make a heuristic: a linear scan through the sorted array will encounter segments in the order `selfpromo`, `sponsor`, and then `intro`, somehow the later `intro` must override the former `sponsor`.

An attempt to create separate entries for each category instead of merging them all together very often results in the creation of milliseconds long chapters. In the linear scan above, `sponsor` ends at 47.68613 while `intro` ends at 48.351925. A separate chapter for `intro` lasts less than a second.

faissaloo replied:
Sounds good

>It is also very hard to make a heuristic  
  
Yeah no kidding lol, took me ages to write the cutting code for Sponskrub to deal with this stuff

nihil-admirari commented:
> Yes, leave sponskrub as-is for now.

Ok, I'll bring it back, and move to [Deprecated Options](https://github.com/yt-dlp/yt-dlp#deprecated-options).

> Yes, please do so if possible

Ok, I'll do.

> It is though.

SponSkrub does in fact implement the cut the way you just described, but the current implementation is different. SponSkrub implementation is easier from the developer perspective,  but it is not necessarily better from the user perspective.

Please take a look at [`test_merge_with_sponsor_chapters_SponsorsWithinChapter`](https://github.com/yt-dlp/yt-dlp/pull/360/files#diff-d1bd28ced589ba16d13310f365004d8d16e534dba6574c642fe516aa41548f2aR102). `expected_no_cut` has the original chapter split: `['c - 1', 's1', 'c - 2', 's2', 'c - 3', 's3', 'c - 4']`, but `expected_cut` does not: `['c']`, only the duration of the chapter has been reduced from 7 to 4.

I would rather want to have that feature present, I don't like to see chapters split for no reason. Please tell me if you dislike that particular feature of cut.

> Nothing needs to change in the chapter merger.

Please take a look at [`Helper.append()`](https://github.com/yt-dlp/yt-dlp/pull/360/files#diff-1051e7d43a7be354ecfc1ddf7411d6f4c14382839df59df5573b3c1d76e67490R279) and [`Helper.handle_intersection()`](https://github.com/yt-dlp/yt-dlp/pull/360/files#diff-1051e7d43a7be354ecfc1ddf7411d6f4c14382839df59df5573b3c1d76e67490R289) inside of [`_merge_with_sponsor_chapters`](https://github.com/yt-dlp/yt-dlp/pull/360/files#diff-1051e7d43a7be354ecfc1ddf7411d6f4c14382839df59df5573b3c1d76e67490R237). Currently its implementation depends on the global `self._cut`. To leave some chapters and cut out others, the flag must be made per chapter.

> In that case, we should split the PP into SponsorBlockPP and RemoveChaptersPP.
> Once we put the modified chapter data in the infodict using SponsorBlock pre-processor, FFmpegMetadataPP will automatically take care of adding these chapters

`FFmpegMetadataPP` does in fact write chapter metadata even in this implementation. [The only changes there](https://github.com/yt-dlp/yt-dlp/pull/360/files?diff=unified&w=1#diff-3e3643b041e66907c439f5f74aca39c2ba6f86547ad4d05c9b1bd3701ce7fbdb) were made to support writing chapters when user requested `--sponsorblock`, but did not request either `--sponsorblock-cut` or `--append-metadata` (you need to hide whitespace changes to see it, most changed lines are due to indentation). I think it's better from UX perspective to make `--sponsorblock` imply `--add-chapters`, instead of forcing people to specify them both. It also matches SponSkrub's implementation. What do you think about it? 

However, it is not the reason why I suggested to extract `ChaptersPP`. These are the parts of the implementation:
1. Modifying chapters object by either splicing sponsor chapters in or by cutting them out.
2. Cut is implemented in a way described above: no unnecessary splits.
3. A new feature that would allow cutting arbitrary chapters, not only sponsors.
4. Cutting the video itself and the subtitles.
5. Writing chapter metadata.

Let's look at ways to combine these pieces together.

`SponsorBlockPP` only queries the API, modifies chapters taking only sponsors into account, and cuts sponsors out of the video. Then `RemoveChaptersPP` has to modify chapters the second time, and cut the video the second time. Chapter modification code has to be extracted to make it available to `RemoveChaptersPP`. `FFmpegMetadataPP` writes chapters to the file as it currently does.

This design tries to separate concerns. The separation of concern comes at the cost of having to modify chapters twice, and having to cut the video twice. At first, that double cut seems avoidable by having `SponsorBlockPP` pass some `info['__sponsor_segments']` down to `RemoveChaptersPP` instead of doing the work itself. Now `RemoveChaptersPP` has to pay attention to SponsorBlock parameters to e.g. implement a cut the way it was described above. That passing of information between post-processors created a tight coupling between them nullifying the separation of concerns that the design tried to achieve in the first place.

Alternative design is to have `ChaptersPP` handling both `--remove-chapters` and SponsorBlock. It contains all the code to splice and cut the chapters (which by necessity implies cutting the video and the subtitles). SponsorBlock code now is just a couple of functions that conditionally attach sponsor chapters to the chapters explicitly requested to be removed. Cutting without splitting is possible since all the information is in the same place.

It is still possible to retain chapter writing in `FFmpegMetadataPP`. But it raises the obvious question: why all chapter handling is in one place, but the actual chapter writing is in another? Moving chapter writing code from `FFmpegMetadataPP` seems like a natural thing to do. It also makes it possible to have `--sponsorblock` imply `--add-chapters` without the necessity to pass an ugly `info['__sponsor_blocked']` down the line. `--add-metadata`  of course would still imply running `ChaptersPP` to match the old behaviour.

All future chapter handling code would have a natural post-processor to tag along to. What do you think about creating `ChaptersPP` in a way described above? Again, nothing changes if chapter writing stays in `FFmpegMetadataPP`. But having separate post-processors for SponsorBlock and `--remove-chapters` doesn't look like a good design due to the reasons described above. And having `--remove-chapters` handled by `SponsorBlockPP` looks rather strange.

faissaloo commented on yt_dlp/postprocessor/sponsorblock.py:
```diff
@@ -0,0 +1,391 @@
+    # See https://ffmpeg.org/general.html#Subtitle-Formats.
+    SUPPORTED_SUBS = 'srt', 'ass', 'vtt'
+
+    def __init__(self, downloader, cut=False, add_keyframes=False, force=False, plaintext=False,
+                 api='https://sponsor.ajay.app', categories=ALL_CATEGORIES):
```
You'll want to also make sure people can choose to keep the modification time of the original file (i.e: `-no-id` functionality that exists in Sponskrub)

nihil-admirari replied:
Modification time is handled by [`real_run_ffmpeg`](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/postprocessor/ffmpeg.py#L244), which I don't touch at all.

[`no-id` from SponSkrub](https://github.com/faissaloo/SponSkrub/blob/master/src/sponskrub/sponskrub.d#L153) is mapped unto [`plaintext`](https://github.com/yt-dlp/yt-dlp/pull/360/files/e5f8f4afa79d37571f6ce76b3b14c135891317cc#diff-1051e7d43a7be354ecfc1ddf7411d6f4c14382839df59df5573b3c1d76e67490R61). Probably I should rename it into `reveal_video_id` to make it clearer.

faissaloo commented on yt_dlp/postprocessor/sponsorblock.py:
```diff
@@ -0,0 +1,391 @@
+        else:  # Custom API endpoint without a protocol. Try HTTPS then fallback to insecure.
+            try:
+                return make_request('https://' + url)
+            except PostProcessingError:
+                return make_request('http://' + url)
```
You should hard error when you can't request over HTTPS and ask the user to specify HTTP if that's what they want, I actually just realised I overlooked this in Sponskrub but you shouldn't do the same because it's bad security practice.

nihil-admirari replied:
OK, I'll remove the `http` branch.

faissaloo reviewed (commented):
Nice work, just a few things that might be worth thinking about

nihil-admirari commented:
@pukkandan, looks like `FFmpegMetadataPP` exists in [the original fork](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/postprocessor/ffmpeg.py#L433), so chapter writing code better be left in there to minimize merge conflicts. But as I said previously
> nothing changes if chapter writing stays in FFmpegMetadataPP.

On the other hand, are you OK with having `ModifyChaptersPP` handling both `--remove-chapters` and SponsorBlock? Creating two tightly coupled post-processors having to communicate through `info[__sponsorblock_segments]` looks like bad design, they better be in the same place to maximize cohesion. And without that communication, implementation of non-splitting cut is impossible.

pukkandan commented:
@nihil-admirari Before I answer your queries, I need a few things clarified. I think it would be easier to discuss this in real time. If you have a discord account, could you hop on to the yt-dlp server

nihil-admirari commented:
Sorry, I don't have a Discord account. Is it OK, if, first, I finish the algorithm rewrite, and then, if you still have any questions, I may try to join Discord?

nihil-admirari commented:
Small changes:
1. Brought SponSkrub back. Its options are hidden from the README, but everything else stays the same. New implementation is disabled by default.
2. Post-processor updates info-json after modifying chapters.
3. Concat implementation is moved to the superclass and is now available to every FFmpeg post-processor.
4. Minimized changes in `FFmpegMetadataPP`. Still `--sponsorblock` implies adding chapters even if `--add-metadata` wasn't explicitly requested.
5. Renamed `plaintext` to `reveal_video_id`.
6. HTTP now must be explicitly requested.
7. Since Python 2 has been deprecated, the code now uses Python >= 3.6 features: `raise ... from ...`, `nonlocal`, format strings, response encoding for JSON data and type annotations. `encodeFilename` is no longer used, `io.open` replaced with plain `open`.

Big change: implementing `--remove-chapters` and `--sponsorblock-cut-categories` turned out to be rather difficult. The algorithm was completely rewritten and is now based on a priority queue.

Please review.

pukkandan commented:
huh, I thought I had already replied to your earlier comment. I must have typed it up and then forgot to submit 😢 Give me some time to collect my thought again on the various points. Then I will review

> README.md update line from Makefile
> generate an enormous diff. Is that line correct?

Yes. The big diff is caused due to the new option names being too long. Anyway, you should revert that and only create the corresponding options for now. We can `make README.md` once we have finalized everything else

nihil-admirari commented:
Reverted changes in README + some small fixes.

Workflow checks failed due to
```
2021-06-07T09:41:51.6758889Z AssertionError: None is not true : Regexp didn't match: '(?:(?:#.*?|\\s*)\\n)*from __future__ import (?:[a-z_]+,\\s*)*unicode_literals' not found, unicode_literals import  missing in /home/runner/work/yt-dlp/yt-dlp/yt_dlp/postprocessor/modify_chapters.py
```

I've added `unicode_literals` import, but I'm not sure that test has any uses now, after all, Python 2 has been deprecated.

pukkandan commented:
> I've added unicode_literals import, but I'm not sure that test has any uses now, after all, Python 2 has been deprecated.

Yes it has no use, I just forgot to disable that test

nihil-admirari commented:
Fixed lint error.

Python 3.4 on Windows fails with `SyntaxError`s, which is expected considering that the code uses 3.6 format strings. Since only Python 3.6 will be supported in the future, can you please just retire Python 3.4?

pukkandan commented:
You can ignore the 3.4 test. By the time this code is merged, it will be removed

phyzical commented:
@nihil-admirari  hey there, not sure if im jumping the gun on giving this a test, but when i try `--sponsorblock --sponsorblock-cut-categories sponsor` for `https://www.youtube.com/watch?v=Dy312cUHumk` i get a `ERROR: Postprocessing: Conversion failed!` and the video playback ends at the first sponsor block.

When i didnt cut i had this warning thrown `Some SponsorBlock segments overlap` incase that helps hint in a direction. i would try to debug it but 

also on my quick usage i wonder if another flag of say `--sponsorblock-cut` should be added which will just cut what is queried so by default that would be all categories, hell id argue we should cut by default if `--sponsorblock` is provided, but i assume this is due to how sponskrub defaults were.

nihil-admirari commented:
> also on my quick usage i wonder if another flag of say `--sponsorblock-cut` should be added which will just cut what is queried so by default that would be all categories

I shortened command line options: dropped `-categories` from `--sponsorblock-query/cut-categories`, and added a shortcut `all` to select everything. Is it better now?

> ... when i try `--sponsorblock` `--sponsorblock-cut-categories sponsor` for `https://www.youtube.com/watch?v=Dy312cUHumk` i get a `ERROR: Postprocessing: Conversion failed!` and the video playback ends at the first sponsor block.

Please always provide `-v` logs in such cases, otherwise I'd have to guess what formats you've been trying to download. MP4 format `-bv*[ext=mp4]+ba[ext=m4a]` worked fine for me.

WEBM format `bv*+ba` threw:
```
Application provided invalid, non monotonically increasing dts to muxer in stream 0: 43417 >= 43410
```
It looks like a rather popular problem. https://ipcamtalk.com/threads/fix-non-monotonous-dts-errors-in-ffmpeg.30465/#post-286226 suggested to add `-copytb 1`, which I did. Unfortunately it didn't work. An additional `-strict experimental` didn't work either, and [FFmpeg documentation](https://ffmpeg.org/ffmpeg-all.html#Codec-Options) warns against using it anyway:
> Note: experimental decoders can pose a security risk, do not use this for decoding untrusted input.

Tested with the latest FFmpeg `master` from https://johnvansickle.com/ffmpeg/.

Unless someone knows a magic workaround, nothing can be done about it. You can try to open an issue in FFmpeg bug tracker, but it looks like there is [a plenty of such bug reports already](https://html.duckduckgo.com/html?q=ffmpeg+non+monotonically+increasing+dts+site:trac.ffmpeg.org). I can only suggest you to switch to MP4, for which cutting does work.

phyzical commented:
@nihil-admirari  thanks for the additions `all` should do the job for what i was suggesting but atm it throws `invalid SponsorBlock category: all`  `--sponsorblock --sponsorblock-cut all`, assume you will need to skip the category check if all is provided

Sorry! i didnt even realize there was a -v i should have looked harder, will in the future 👍 .

sorry, ive jsut come from the base yt-dl and my format settings were not working with yt-dlp, so i was just using whatever the default detected was. i can confirm it removes the sponsor as expected with the format you suggested now and with almost no overhead to the processing 😄 good job! once the all category issue is resolved i will give it a test with subtitles also.

Though it should be noted that others may run into similar issues as i did, maybe a note in the readme for sponsor block should be added warning about format issues once that is done? though like you've found it looks more like an ffmpeg issue than an issue with your post processor

pukkandan commented:
> Application provided invalid, non monotonically increasing dts to muxer in stream 0: 43417 >= 43410

This issue does not exist in ffmpeg versions <= `3.0.1`. See https://github.com/ytdl-org/youtube-dl/issues/28042

nihil-admirari commented:
> atm it throws `invalid SponsorBlock category: all`

Have you pulled the latest changes from this branch? Caveat: because of force push (rebase on `master`), you may need to remove you local branch, then pull, and then checkout this branch again.

> When i didnt cut i had this warning thrown `Some SponsorBlock segments overlap`

Actually, it was supposed to warn even when doing a cut. Fixed.

> should be added warning about format issues once that is done

Added a warning.

I tried various FFmpeg options that were mentioned as workarounds in various places.
1. `+discardcorrupt`, `+genpts`, `+igndts` fail with error.
2. `-bsf setts=dts=PREV_INDTS/PREV_OUTDTS` and various other combinations fail too.
3. `-bsf vp9_superframe/vp9_superframe_split/vp9_raw_reorder` all fail.
4. `-vsync drop` proceeds to completion, but the resulting video is broken: VLC only shows static frames that replace each other infrequently.
5. `+noparse +nofillin` doubles the file size, but the resulting video is pure black screen.
6. Forced re-encoding with `--remove-chapters-force-keyframes` works (at least on the smallest video `-S +size`), but is terribly slow.

phyzical commented:
@nihil-admirari not sure what happened as i had the new flags, in any case `all` now works, tested one with subtitles looks to keep in sync. from my few test runs all looks good to me 👍

nihil-admirari commented:
> we should cut by default if `--sponsorblock` is provided, but i assume this is due to how sponskrub defaults were.

Indeed, cutting was disabled by default to match SponSkrub behaviour. It made sense for SponSkrub, which implements cut by re-encoding. In this implementation cutting is fast, and, I presume, the majority of people who use SponsorBlock would like to cut rather than contemplate the chapters. Changed to cut by default.

nihil-admirari commented:
@phyzical, if you still feel like testing, you can find patched FFmpeg here: https://github.com/nihil-admirari/FFmpeg-With-VP9-Timestamp-Fix/releases. It should handle VP9 videos with corrupted timestamps.

phyzical commented:
@nihil-admirari no worries ill give it a crack later today after work 👍

phyzical commented:
i'm not actually sure how to build the patched ffmpeg, when i try the build script it just kept failing `#9 1.936 + shopt -s globstar
#9 1.937 /bin/sh: ./build.sh: line 3: shopt: not found`

nihil-admirari commented:
@phyzical, you don't need to build anything, [releases' page](https://github.com/nihil-admirari/FFmpeg-With-VP9-Timestamp-Fix/releases) already has the binaries. `ffmpeg-N-*` are builds of FFmpeg `master`, `ffmpeg-n4.4-20-*` are builds of the latest release branch. There are `win32`, `win64`, and `linux64` binaries for each variant. Sorry, if my previous comment mislead you.

phyzical commented:
@nihil-admirari thanks for the guidance, but i must still be a bit too ignorant, when i try to use the linux binary i get `ERROR: Preprocessing: ffmpeg not found. Please install or provide the path using --ffmpeg-location`. to note im trying on an alpine based docker image.

i am copying across the binary to `/tmp/ffmpeg` then providing it via ` --ffmpeg-location /tmp/ffmpeg` to yt-dl command but seems to have no affect?

nihil-admirari commented:
@phyzical, a couple of questions:
1. In `--ffmpeg-location /tmp/ffmpeg` is `/tmp/ffmpeg` a directory containing `ffmpeg` and `ffprobe` (you need that one too), or the `ffmpeg` binary itself? It must be a directory containing binaries.
2. Just to make sure: have executable bit been preserved when copying to `/tmp`?
3. Just to make sure: is `/tmp` mounted `exec`?
4. Can you try a different location? Mine `ffmpeg` is in `--ffmpeg-location /opt/ffmpeg-n4.4-20-g2131bc585c-linux64-gpl-4.4/bin/`.

If still nothing works:
> to note im trying on an alpine based docker image.

Alpine uses musl. Provided binaries should be statically linked against glibc and should work on musl distros, but just in case they don't, could you please retest with a glibc-based distro?

phyzical commented:
@nihil-admirari 
thanks for the reply,
1. ah yep i was only specifying the ffmpeg directly, but still didnt change the result sadly
2. yeah 100% set
![image](https://user-images.githubusercontent.com/5182053/122671681-70d57a80-d1fa-11eb-8f37-a86c99783d22.png)
3. yeah its on there
![image](https://user-images.githubusercontent.com/5182053/122671592-07ee0280-d1fa-11eb-8121-e6877d392b0d.png)
4. tried /opt/ffmpeg still no luck

But when i tried an ubuntu image sure enough it works, so something with the alpine image was going weird 
And can confirm the custom ffmpeg fixes the issue with default formats when overlapping segments

nihil-admirari commented:
> can confirm the custom ffmpeg fixes the issue with default formats

That's great!

> But when i tried an ubuntu image sure enough it works, so something with the alpine image was going weird

Here's FFmpeg-Builds author's [comments](https://github.com/BtbN/FFmpeg-Builds/issues/55#issuecomment-839936548) (emphasis mine):

> I've been looking into that the last couple days actually, and it's surprisingly painful.
Main issue is **glibcs inability to actually be linked statically**, so you end up with static builds that only work on super bleeding edge systems that **use the same or newer of a glibc than your build system**.

Alpine's wiki actually has a [glibc entry](https://wiki.alpinelinux.org/wiki/Running_glibc_programs). The only sane option (messing with `chroot` is pointless when you can simply use a different Docker image) is to try to install a compatibility layer: `apk add gcompat`. If it still does not work, the only option is to ask [FFmpeg-Builds](https://github.com/BtbN/FFmpeg-Builds) to provide musl-linked binaries: I've no idea how to actually build FFmpeg, my fork is just a trivial patch applied to what is provided by the upstream.

nihil-admirari commented:
@pukkandan, do you have any questions about this pull request? Can it be merged?

nihil-admirari commented:
1. Added support for preview category (see https://github.com/ajayyy/SponsorBlock/issues/444).
2. Removed `copytb` from concat options: it was useless in fixing FFmpeg DTS issue anyway.
3. To avoid resolving an enormous merge conflict in `ffmpeg.py` multiple times, all commits have been squashed into one.

pukkandan commented:
hey, sorry that this has been delayed so much. So many other issues have been popping up recently and finalizing this is quite a big time commitment on my part.

I promise I'll make a priority to get this merged as soon as I am done with the upcoming release (a couple days at most)

thanks for sticking around despite everything

Sipherdrakon commented:
so I was just curious if any of the changes here will allow for auto-merging playlists like mtv, vh1, etc.?

pukkandan commented:
no, not directly. This does implement a concat demuxer which we can use to implement that functionality

Sipherdrakon commented:
Awesome, that's what I thought when I saw the ffmpeg postprocessor changes. Cool, thank you for the information. :-)

pukkandan commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -34,23 +33,13 @@ def run(self, info):
-                    f'Skipping {self.pp_key()} since the video was already downloaded. '
-                    'Use --force-remove-chapters to remove the chapters anyway')
-                return [], info
+        if not info.get('__real_download') and abs(duration - info['duration']) > 1:
+            self.report_warning(f'Skipping {self.pp_key()} since the video appears to be already cut.')
+            return [], info
```
I think we can improve this

pseudocode
```
if real_duration != duration:
    if duration == duration_after_cut:
        to_screen(Already cut)
        return
    else if real_download:
        warn(Durations mismatch)    # Since we just downloaded it, just assume extracted duration is wrong. So proceed to cut anyway
    else:
       error(Durations mismatch.)  # Since addchapters will now break
 ```

nihil-admirari replied:
Implemented but without the `warn(Durations mismatch)`. Warning would be a result of extractor duration inaccuracy. Doesn't look like there is something that can be done with it, so no point to annoy the user.

pukkandan commented on yt_dlp/__init__.py:
```diff
@@ -488,13 +492,12 @@ def report_conflict(arg1, arg2):
         except re.error as err:
             parser.error(f'invalid --remove-chapters regex {regex!r} - {err}')
-    if opts.remove_chapters or opts.remove_sponsor_segments:
+    if opts.remove_chapters or opts.sponsorblock_remove:
```
this should run for marking as well

nihil-admirari replied:
Fixed.

pukkandan commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -34,23 +33,13 @@ def run(self, info):
-        # TODO: Refactor _remove_marked_arrange_sponsors to not need this
-        for c in sponsor_chapters:
-            c['categories'] = [(c['category'], c['start_time'], c['end_time'])]
-
```
Moving this down breaks the code. Try running `yt-dlp fFFSKedy9f4 --sponsorblock-mark all`

nihil-admirari replied:
Moved this part to `SponsorBlockPP`.

nihil-admirari commented on yt_dlp/__init__.py:
```diff
@@ -352,21 +352,26 @@ def metadataparser_actions(f):
         warnings.append('%s is ignored since %s was given' % (arg2, arg1))
 
-    if (opts.remove_chapters or opts.remove_sponsor_segments) and opts.sponskrub is not False:
+    if (opts.remove_chapters or sponsorblock_query) and opts.sponskrub is not False:
```
No, remove with SponSkrub will result in double remove.

pukkandan replied:
yeah, idk what I was thinking while writing this

nihil-admirari commented on yt_dlp/__init__.py:
```diff
@@ -352,21 +352,26 @@ def metadataparser_actions(f):
-                report_conflict('--remove-sponsor-segments', '--sponskrub')
             if opts.remove_chapters:
                 report_conflict('--remove-chapters', '--sponskrub')
+            if opts.sponsorblock_mark:
+                report_conflict('--sponsorblock-mark', '--sponskrub')
```
1. Why would anyone wanna do that?
2. SponSkrub marks are not the same since it uses a different algorithm.

pukkandan replied:
I said "technically" it doesnt conflict. tbh, I dont really care if whether or not we consider it a conflict

pukkandan commented:
> > But not having consistent names for the chapters makes this use-case impossible. This issue exists for the overlap handling as well

> Sponsor chapters have [SponsorBlock]: prefix that never changes. Is it enough to set up a skip, or the part after the colon also matters?

Both in potplayer, and mkv's chapter-skip, the exact chapter name should be given. Not sure how it is in other players

---

> Overlapped segments result in a large chapter with a combined title: Intro/Sponsor/Self-Promotion etc, even if overlap is small. Tiny chapters can be created only by cuts. They can be ignored by changing append_chapter slightly.

Oh, seems I misunderstood how the overlaps were being handled. But this is even worse imo

Take this example:
```
00-20 = A
05-10 = B
25-35 = C
30-50 = D
49-60 = E
```

What current implementation gives:
```
00-20 = A/B
25-60 = C/D/E
```

This is throwing away a lot of details, which is bad, especially since adjacent chapters are always **supposed to** overlap slightly. If we want to avoid any information loss, it should instead be:
```
00-05 = A
05-10 = A/B
10-20 = A
25-30 = C
30-35 = C/D
35-49 = D
49-50 = D/E
50-60 = E
```
(this is what I thought u were doing)

Mainly we have to think of 3 kinds of overlap.
1. The slight overlap that is expected b/w adjacent chapters (D/E) - For this, having a seperate `D/E` chapter is undesirable and the overlap section should be considered part of one or the other
2. A significant overlap b/w adjacent chapters (C/D) - If the data in the API was submittted correctly, this should never happen (except for non-music). However, this does in fact sometimes happen. In this case, either `C/D` or just either `C` or `D` are reasonable. If not using using `C/D`, and the chapters differ significantly in length, it makes intuitive sense to use the name of the shorter chapter
3. A chapter inside another (A/B) - For this, the inner chapter can be named either `A/B` or just `B`

A simple system that can all these situations is to simply **consider overlaps as part of the smaller chapter**. With that, the results would be:
```
00-05 = A
05-10 = B
10-20 = A
25-35 = C
35-49 = D
49-60 = E
```

Another method could be to only do this if the overlap is small (say <5sec) and use the combined chapter names for the rest. That would look like:
```
00-05 = A
05-10 = A/B
10-20 = A
25-30 = C
30-35 = C/D
35-49 = D
50-60 = E
```
This still does however have the issue with skipping in players that I mentioned

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,326 @@
+            if not sub_file or not os.path.exists(sub_file):
+                continue
+            ext = sub['ext']
+            if ext not in FFmpegSubtitlesConvertorPP.SUPPORTED_EXTS:
```
Reverted to `FFmpegSubtitlesConvertorPP.SUPPORTED_EXTS`.

nihil-admirari commented on yt_dlp/options.py:
```diff
@@ -1390,38 +1384,27 @@ def _dict_from_options_callback(
 
     sponsorblock = optparse.OptionGroup(parser, 'SponsorBlock Options')
```
Brought section description back.

nihil-admirari commented on yt_dlp/options.py:
```diff
@@ -1390,38 +1384,27 @@ def _dict_from_options_callback(
         help=(
-            'Sponsorblock categories to query SponsorBlock API for, seperated by commas. '
+            'SponsorBlock categories to create chapters for, separated by commas. '
             'See https://github.com/ajayyy/SponsorBlock/wiki/Segment-Categories'
```
Updated the link.

nihil-admirari commented on yt_dlp/options.py:
```diff
@@ -1390,38 +1384,27 @@ def _dict_from_options_callback(
+        help=(
+            'SponsorBlock categories to be removed from the video file, separated by commas. '
+            'If a category is present in both mark and remove, remove takes precedence. '
+            'Available categories are all, %s. You can prefix the category with a "-" to exempt it from being removed. '
```
Did.

nihil-admirari commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -342,7 +342,9 @@ def _concat_spec(cls, in_files, concat_opts=None):
         for file, opts in zip(in_files, concat_opts):
             yield f'file {cls._quote_for_ffmpeg(cls._ffmpeg_filename_argument(file))}\n'
-            yield from (f'{key} {val}\n' for key, val in opts.items())
+            for directive in 'inpoint', 'outpoint', 'duration':
```
Added a comment.

nyuszika7h commented:
@nihil-admirari

> Alteration handles it as well: `CASE|((?i)no-case)`.

Nope, that doesn't work.

```
>>> re.search(r'CASE|((?i)no-case)', 'case')
<stdin>:1: DeprecationWarning: Flags not at the start of the expression 'CASE|((?i)no-case)'
<re.Match object; span=(0, 4), match='case'>
```

nihil-admirari commented:
> Nope, that doesn't work:

My bad, I forgot regex syntax. `CASE|(?i:no-case)` works.

```
re.search(r'CASE|(?i:no-case)', 'case')  # Nothing
```

```
re.search(r'CASE|(?i:no-case)', 'No-CaSe')
<re.Match object; span=(0, 7), match='No-CaSe'>
```

nihil-admirari commented:
Is it really necessary to handle sponsor overlaps differently? Previous implementation required the re-encoding of an entire video, and that's why people used chapter mode, but now sponsors can be removed hassle-free: the primary use case for chapters is gone.

> Both in potplayer, and mkv's chapter-skip, the exact chapter name should be given. Not sure how it is in other players

> This still does however have the issue with skipping in players that I mentioned

Maybe PotPlayer devs can implement a skip by regex?

It's rather hard to correctly handle overlaps in all cases, differently from the way they are currently handled.

> A simple system that can all these situations is to simply consider overlaps as part of the smaller chapter.

More than two sponsors can overlap simultaneously. How ties are to be broken?

pukkandan commented:
> Is it really necessary to handle sponsor overlaps differently? Previous implementation required the re-encoding of an entire video, and that's why people used chapter mode, but now sponsors can be removed hassle-free: the primary use case for chapters is gone.

I don't know about others, but the chapter feature of sponskrub was actually requested originally by me because I do not want any destructive operation on the videos I archive. If I want to later see the sponsor section for whatever reason, or if it was marked incorrectly, I want the option to be able to get back the data without redownloading

> It's rather hard to correctly handle overlaps in all cases, differently from the way they are currently handled.

> > A simple system that can all these situations is to simply consider overlaps as part of the smaller chapter.

> More than two sponsors can overlap simultaneously. How ties are to be broken?

I know how to implement this easily with multiple passes. I am not entirely sure how to do it in a single pass. Let me have a go at it

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,326 @@
+                raise PostProcessingError('ffprobe returned empty duration')
+            return duration
+        except PostProcessingError as err:
+            raise PostProcessingError(f'Cannot determine the video duration; {err}')
```
I haven't thought that Python would print `During handling of the above exception, another exception occurred`, I thought it's gonna be lost. Anyway, exceptions are usually chained with `raise ... from e`, which will produce `The above exception was the direct cause of the following exception`.

Message of an inner exception is obscure because it's a test exception, not a real one. The real exception would have a better description.

nihil-admirari commented:
Implemented sponsor splits:
```
00-05 = A
05-10 = A/B
10-20 = A
25-30 = C
30-35 = C/D
35-49 = D
49-50 = D/E
50-60 = E
```
Tiny chapters (< 1 sec.) are appended to the previous chapter.

nihil-admirari commented:
> This still does however have the issue with skipping in players that I mentioned

Added a template for sponsor chapter names. I don't really understand the machinery of output template, and thus currently it's using `str.format`.

pukkandan commented:
> Added a template for sponsor chapter names

This was the perfect idea. I have modified it to use output template instead. Then, added code to merge adjacent chapters based on title rather than category. This now satisfies both our uses. By default, no data is lost in the chapter names, but by using `--sponsorblock-chapter-title "SponsorBlock: %(name)s"`, I can get the chapters to use only the title of the smallest category

I've also made a few small changes. Have a look at the commit messages

If you are satisfied with how it is now, I believe this can be merged

pukkandan commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -223,43 +251,59 @@ def append_chapter(c):
+                })
+                outtmpl, tmpl_dict = self._downloader.prepare_outtmpl(self._sponsorblock_chapter_title, c)
+                c['title'] = self._downloader.escape_outtmpl(outtmpl) % tmpl_dict
+            if i > 0 and c['title'] == new_chapters[i - 1]['title']:
+                new_chapters[i - 1]['end_time'] = c['end_time']
+                new_chapters.pop(i)
+                i -= 1
```
This assumes that `new_chapters` is already sorted. This is always correct, no?

nihil-admirari replied:
Yes.

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,333 @@
+            if not info.get('__real_download'):
+                raise PostProcessingError('Cannot cut video since the real and expected durations mismatch. '
+                                          'Different chapters may have already been removed')
+                return [], info
```
Unreachable code.

nihil-admirari replied:
Fixed.

pukkandan replied:
oh, right. I added the return for `report_error`. But now it is unreachable

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,333 @@
+)
+
+
+_TINY_SPONSOR_OVERLAP_DURATION = 1
```
Affects ordinary chapters too. `TINY_CHAPTER_DURATION` was a better name.

nihil-admirari replied:
Brought back old name.

pukkandan replied:
I thought only sponsor-sponsor overlaps were affected by this

nihil-admirari replied:
No. Tiny chapters present in the original list are always preserved, but if a cut or a sponsor overlap result in a creation of a tiny chapter out of a normal one, that tiny chapter will also be ignored.

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,333 @@
+                c['title'] = self._downloader.escape_outtmpl(outtmpl) % tmpl_dict
+            if i > 0 and c['title'] == new_chapters[i - 1]['title']:
+                new_chapters[i - 1]['end_time'] = c['end_time']
+                new_chapters.pop(i)
```
Better to create a new list rather than popping the old one.

nihil-admirari replied:
Fixed by creating a new list instead.

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,333 @@
+            c.pop('_was_cut', None)
+            cats = c.pop('_categories', None)
+            if cats:
+                category = min(cats, key=lambda c: c[2] - c[1])[0]
```
Why smallest category? (Largest would've been less surprising).

pukkandan replied:
Because there ae instances with a sponsor section inside another. For example, a sponsor inside an outro. In this case, it makes sense to have 2 outro chapters around a sponsor chapter

nihil-admirari replied:
There are also overlapping sponsors, for which this might not be an appropriate heuristic. Plus it's not really the smallest segment: it's the smallest segment before cuts were taken into account. After the cut, a formerly largest sponsor may actually become the smallest.

If people from discord prefer it that way, let's leave it as is.

pukkandan replied:
> There are also overlapping sponsors, for which this might not be an appropriate heuristic

I agree, but this is good enough for my own use. We'll look into improving it if/when someone has a better idea. Also, considering we keep the combined name by default, I don't think it is a huge deal

nihil-admirari commented on yt_dlp/postprocessor/sponsorblock.py:
```diff
@@ -0,0 +1,96 @@
+        'sponsor': 'Sponsor',
+        'intro': 'Intermission/Intro Animation',
+        'outro': 'Endcards/Credits',
+        'selfpromo': 'Unpaid/Self Promotion',
```
IMHO, old names were better.

pukkandan replied:
I personally don't care either way, but these are the actual names as documented here: https://wiki.sponsor.ajay.app/index.php/Segment_Categories

nihil-admirari commented on yt_dlp/postprocessor/modify_chapters.py:
```diff
@@ -0,0 +1,333 @@
+                outtmpl, tmpl_dict = self._downloader.prepare_outtmpl(self._sponsorblock_chapter_title, c)
+                c['title'] = self._downloader.escape_outtmpl(outtmpl) % tmpl_dict
+                # Merge identically named sponsors.
+                if (named_new_chapters and 'categories' in named_new_chapters[-1]
```
This version merges only sponsor. Identically named original chapters are unaffected.

nihil-admirari commented:
> If you are satisfied with how it is now, I believe this can be merged

I made some minor fixes. I'm OK with merging this, though I don't quite understand why the smallest sponsor contributing to overlap is the most representative of the overlap itself.

nihil-admirari commented on yt_dlp/options.py:
```diff
@@ -1354,41 +1364,90 @@ def _dict_from_options_callback(
+        help='Remove chapters whose title matches the given regular expression. This option can be used multiple times')
+    postproc.add_option(
+        '--no-remove-chapters', dest='remove_chapters', action='store_const', const=None,
+        help='Do not remove any normal chapters from the file (default)')
```
Better to remove `normal`. From the user perspective it's not clear which chapters aren't normal. This distinction is internal and stems from the fact that remove chapters and remove sponsors are handled by the same code.

pukkandan replied:
Since sponsor sections are also marked as chapter when marking, I thought this might cause confusion

pukkandan replied:
```suggestion
        help='Do not remove any chapters from the file (default)')
```

pukkandan commented:
@nihil-admirari Thanks for sticking with me till the end of the PR even though it took some large delays and quite a few detours to get this done 😄
